src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -231,110 +327,28 @@ std::string CHLSLCompiler::preprocessShader(std::string&& code, IShader::E_SHADE
 {
     using lex_token_t = boost::wave::cpplexer::lex_token<>;
     using lex_iterator_t = boost::wave::cpplexer::lex_iterator<lex_token_t>;
-    using wave_context_t = boost::wave::context<
-        core::string::iterator,
-        lex_iterator_t,
-        boost::wave::iteration_context_policies::load_file_to_string/*,
-        TODO: OurCustomDirectiveHooks -> for pragmas and includes!
-        */
-    >;
-
-    // TODO: change `code` to `const core::string&` because its supposed to be immutable
-    wave_context_t context(code.begin(),code.end(),preprocessOptions.sourceIdentifier.data()/*,TODO: instance of OurCustomDirectiveHooks*/);
-//    context.add_include_path
-//    context.add_sysinclude_path <- for dem builtins! / preprocessOptions.includeFinder?
-//    context.add_macro_definition from preprocessOptions.extraDefines
-    core::string resolvedString;
-    // TODO: fill `resolvedString` with `[context.begin(),context.end()]`
-
-    // Line 1 comes before all the extra defines in the main shader
-    insertIntoStart(code, std::ostringstream(std::string(IShaderCompiler::PREPROC_DIRECTIVE_ENABLER) + "line 1\n"));
-
-    uint32_t defineLeadingLinesMain = 1;
-    uint32_t leadingLinesImports = IShaderCompiler::encloseWithinExtraInclGuardsLeadingLines(preprocessOptions.maxSelfInclusionCount + 1u);
-    if (preprocessOptions.extraDefines.size())
-    {
-        insertExtraDefines(code, preprocessOptions.extraDefines);
-        defineLeadingLinesMain += preprocessOptions.extraDefines.size();
+    using wave_context_t = boost::wave::context<core::string::iterator, lex_iterator_t, hlsl::impl::load_file_or_builtin_to_string, hlsl::impl::custom_preprocessing_hooks>;
+
+    hlsl::impl::custom_preprocessing_hooks hooks(preprocessOptions, stage);
+    std::string startingFileIdentifier = std::string("../") + preprocessOptions.sourceIdentifier.data();
+    wave_context_t context(code.begin(), code.end(), startingFileIdentifier.data(), hooks);
+
+    //TODO fix bad syntax and uncomment
+    // instead of defining extraDefines as "NBL_GLSL_LIMIT_MAX_IMAGE_DIMENSION_1D 32768", 
+    // now define them as "NBL_GLSL_LIMIT_MAX_IMAGE_DIMENSION_1D=32768" 
+    // to match boost wave syntax
+    // https://www.boost.org/doc/libs/1_82_0/libs/wave/doc/class_reference_context.html#:~:text=Maintain%20defined%20macros-,add_macro_definition,-bool%20add_macro_definition
```

Comment:
change the api of `preprocessOptions.extraDefines` to be a signature/definition "pair"
```cpp
struct MacroDefinition
{
    std::string_view identifier;
    std::string_view definition;
};
```

then you can slap `=` between `identifier` and `definition`